src/intrinsics.cpp

```diff
@@ -432,6 +450,19 @@ static jl_cgval_t emit_runtime_call(jl_codectx_t &ctx, JL_I::intrinsic f, const
     return mark_julia_type(ctx, r, true, (jl_value_t*)jl_any_type);
 }
 
+static jl_cgval_t emit_runtime_intrinsic(jl_codectx_t &ctx, jl_cgval_t *f, const jl_cgval_t *argv, size_t nargs)
```

Comment:
The function you need to codegen here precisely is `jl_f_intrinsic_call`. I am not sure how much can be done to make that dispatch faster here, but you could likely generate N copies of that function (in C) with varying numbers of arguments. Though the cost seems to be smeared across a lot of other places than that:
```
julia> using Profile, BenchmarkTools
julia> @bprofile $(Ref(Core.Intrinsics.add_int))[](1, 2)
julia> D = Profile.retrieve();
julia> Profile.print(D..., C=true)
...
 14╎ 710  /data/vtjnash/julia/src/gf.c:2348; _jl_invoke
 15╎  15   /data/vtjnash/julia/src/builtins.c:1745; jl_f_intrinsic_call
 16╎  16   /data/vtjnash/julia/src/builtins.c:1749; jl_f_intrinsic_call
 16╎  16   /data/vtjnash/julia/src/builtins.c:1752; jl_f_intrinsic_call
 35╎  35   /data/vtjnash/julia/src/builtins.c:1763; jl_f_intrinsic_call
 11╎  11   /data/vtjnash/julia/src/gf.c:2124; jl_fptr_args
 14╎  14   /data/vtjnash/julia/src/runtime_intrinsics.c:1152; jl_add_int
 32╎  32   /data/vtjnash/julia/src/runtime_intrinsics.c:969; jl_iintrinsic_2
 19╎  19   /data/vtjnash/julia/src/runtime_intrinsics.c:974; jl_iintrinsic_2
  3╎  3    /data/vtjnash/julia/src/runtime_intrinsics.c:975; jl_iintrinsic_2
  6╎  6    /data/vtjnash/julia/src/runtime_intrinsics.c:976; jl_iintrinsic_2
   ╎  2    /data/vtjnash/julia/src/runtime_intrinsics.c:979; jl_iintrinsic_2
  2╎   2    /data/vtjnash/julia/src/julia.h:1251; jl_is_primitivetype
   ╎  44   /data/vtjnash/julia/src/runtime_intrinsics.c:982; jl_iintrinsic_2
 29╎   29   /data/vtjnash/julia/src/julia.h:1249; jl_is_primitivetype
  6╎   6    /data/vtjnash/julia/src/julia.h:1250; jl_is_primitivetype
  3╎   3    /data/vtjnash/julia/src/julia.h:1251; jl_is_primitivetype
  6╎   6    /data/vtjnash/julia/src/julia.h:1252; jl_is_primitivetype
   ╎  61   /data/vtjnash/julia/src/runtime_intrinsics.c:986; jl_iintrinsic_2
 14╎   14   /data/vtjnash/julia/src/support/dtypes.h:192; next_power_of_two
  9╎   9    /data/vtjnash/julia/src/support/dtypes.h:193; next_power_of_two
 10╎   10   /data/vtjnash/julia/src/support/dtypes.h:194; next_power_of_two
  8╎   8    /data/vtjnash/julia/src/support/dtypes.h:195; next_power_of_two
 20╎   20   /data/vtjnash/julia/src/support/dtypes.h:196; next_power_of_two
  6╎  6    /data/vtjnash/julia/src/runtime_intrinsics.c:987; jl_iintrinsic_2
 27╎  371  /data/vtjnash/julia/src/runtime_intrinsics.c:1004; jl_iintrinsic_2
  8╎   8    /data/vtjnash/julia/src/datatype.c:801; ijl_new_bits
  1╎   1    /data/vtjnash/julia/src/runtime_intrinsics.c:1152; jl_add_int64
 36╎   36   /data/vtjnash/julia/src/runtime_intrinsics.c:1009; jl_intrinsiclambda_2
122╎   122  /data/vtjnash/julia/src/runtime_intrinsics.c:1010; jl_intrinsiclambda_2
   ╎   5    /data/vtjnash/julia/src/runtime_intrinsics.c:1011; jl_intrinsiclambda_2
  5╎    5    /data/vtjnash/julia/src/runtime_intrinsics.c:894; select_intrinsic_2
  8╎   23   /data/vtjnash/julia/src/runtime_intrinsics.c:1012; jl_intrinsiclambda_2
 15╎    15   /data/vtjnash/julia/src/runtime_intrinsics.c:1152; jl_add_int64
 11╎   124  /data/vtjnash/julia/src/runtime_intrinsics.c:1013; jl_intrinsiclambda_2
 11╎    11   /data/vtjnash/julia/src/datatype.c:1181; ijl_box_int64
 16╎    16   /data/vtjnash/julia/src/datatype.c:801; ijl_new_bits
 24╎    24   /data/vtjnash/julia/src/datatype.c:804; ijl_new_bits
  1╎    1    /data/vtjnash/julia/src/datatype.c:806; ijl_new_bits
 12╎    12   /data/vtjnash/julia/src/datatype.c:809; ijl_new_bits
 21╎    21   /data/vtjnash/julia/src/datatype.c:810; ijl_new_bits
 28╎    28   /data/vtjnash/julia/src/datatype.c:824; ijl_new_bits
 25╎   25   /data/vtjnash/julia/src/runtime_intrinsics.c:1014; jl_intrinsiclambda_2
 44╎  44   /data/vtjnash/julia/src/runtime_intrinsics.c:1006; jl_iintrinsic_2
  1╎  1    /data/vtjnash/julia/src/runtime_intrinsics.c:1014; jl_intrinsiclambda_2
...
```